File: test_suites/common/fixture.py
#!/usr/bin/env python3
"""从一个固定规则生成三组共用的 3000 Node 静态、动态和指标数据。"""

from __future__ import annotations

import hashlib
import json
from pathlib import Path
from typing import Any

from .runtime import write_json


METRIC_NAMES = (
    "cpuUsageRatio", "memoryUsageRatio",
    "networkReceiveBytesPerSecond", "networkTransmitBytesPerSecond",
    "networkReceivePacketsPerSecond", "networkTransmitPacketsPerSecond",
    "networkReceiveDropRatio", "networkTransmitDropRatio",
    "networkReceiveErrorRatio", "networkTransmitErrorRatio",
    "tcpRetransmitRatio", "networkLinkUpRatio", "networkUtilizationRatio",
    "availableBandwidthBytesPerSecond",
)


def canonical_hash(value: Any) -> str:
    raw = json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(",", ":")).encode()
    return "sha256:" + hashlib.sha256(raw).hexdigest()


def load_config(path: Path | None = None) -> dict[str, Any]:
    source = path or Path(__file__).with_name("fixture-config.json")
    return json.loads(source.read_text(encoding="utf-8"))


def generate(config: dict[str, Any]) -> dict[str, Any]:
    count = int(config["nodeCount"])
    topology_config = config["topology"]
    nodes_per_leaf = int(topology_config["nodesPerLeaf"])
    leaf_count = int(topology_config["leafCount"])
    border_count = int(topology_config["borderCount"])
    leaves_per_core = leaf_count // int(topology_config["coreCount"])
    nodes: list[dict[str, Any]] = []
    states: list[dict[str, Any]] = []
    metrics: dict[str, dict[str, float]] = {}
    kubernetes_nodes: list[dict[str, Any]] = []
    kubernetes_pods: list[dict[str, Any]] = []

    for number in range(1, count + 1):
        index = number - 1
        leaf_number = index // nodes_per_leaf + 1
        core_number = (leaf_number - 1) // leaves_per_core + 1
        # 150 Leaf 均匀挂到 4 个 Border：38、37、38、37。
        border_number = min(border_count, (leaf_number - 1) * border_count // leaf_count + 1)
        node_name = f"worker-{number:04d}"
        uid = f"uid-worker-{number:04d}"
        leaf = f"leaf-{leaf_number:03d}"
        border = f"border-{border_number:02d}"
        core = f"core-{core_number:02d}"
        bandwidth = float((25, 40, 50, 100)[(leaf_number - 1) % 4])
        latency = float((4.0, 2.5, 1.5, 0.8)[(leaf_number - 1) % 4])
        labels = {
            "tests.ngg.io/worker": "true",
            "topology.demo.ngg.io/topology-version": "dc-core-border-leaf-v1",
            "topology.demo.ngg.io/core-switch": core,
            "topology.demo.ngg.io/border-switch": border,
            "topology.demo.ngg.io/leaf-switch": leaf,
            "topology.demo.ngg.io/bandwidth-gbps": str(bandwidth),
            "topology.demo.ngg.io/latency-ms": str(latency),
        }
        topology = {
            "dataCenter": topology_config["dataCenter"],
            "coreSwitchId": core,
            "borderSwitchId": border,
            "leafSwitchId": leaf,
            "bandwidthGbps": bandwidth,
            "latencyMillis": latency,
        }
        node = {
            "nodeName": node_name,
            "nodeUID": uid,
            "createdAt": "2026-08-19T00:00:00Z",
            "allocatable": {"cpu": "32", "memory": "128Gi", "nvidia.com/gpu": "4"},
            "labels": labels,
            "topology": topology,
        }
        nodes.append(node)
        in_use = number % int(config["dynamicState"]["inUseEveryNthNode"]) == 0
        states.append({"nodeUID": uid, "inUse": in_use})
        cpu = round(0.08 + ((leaf_number * 13 + number * 7) % 65) / 100, 4)
        memory = round(0.10 + ((leaf_number * 11 + number * 5) % 60) / 100, 4)
        receive_bps = float(20_000_000 + (number * 7919) % 600_000_000)
        transmit_bps = float(15_000_000 + (number * 6151) % 500_000_000)
        receive_pps = float(20_000 + (number * 97) % 800_000)
        transmit_pps = float(18_000 + (number * 89) % 700_000)
        utilization = min(0.95, (receive_bps + transmit_bps) / (bandwidth * 125_000_000))
        metrics[node_name] = {
            "cpuUsageRatio": min(cpu, 0.95),
            "memoryUsageRatio": min(memory, 0.95),
            "networkReceiveBytesPerSecond": receive_bps,
            "networkTransmitBytesPerSecond": transmit_bps,
            "networkReceivePacketsPerSecond": receive_pps,
            "networkTransmitPacketsPerSecond": transmit_pps,
            "networkReceiveDropRatio": round((number % 11) / 10000, 6),
            "networkTransmitDropRatio": round((number % 13) / 10000, 6),
            "networkReceiveErrorRatio": round((number % 7) / 20000, 6),
            "networkTransmitErrorRatio": round((number % 5) / 20000, 6),
            "tcpRetransmitRatio": round((number % 17) / 10000, 6),
            "networkLinkUpRatio": 1.0 if number % 101 else 0.5,
            "networkUtilizationRatio": round(utilization, 6),
            "availableBandwidthBytesPerSecond": max(0.0, bandwidth * 125_000_000 - receive_bps - transmit_bps),
        }
        kubernetes_nodes.append({
            "apiVersion": "v1", "kind": "Node",
            "metadata": {"name": node_name, "uid": uid, "labels": labels},
            "spec": {"unschedulable": in_use},
            "status": {
                "allocatable": node["allocatable"],
                "capacity": node["allocatable"],
                "conditions": [{"type": "Ready", "status": "True", "reason": "MockReady", "message": "fixture node"}],
            },
        })
        # 约 100 个已绑定 Pod 提供真实的 requestedResources 聚合输入；
        # 它们只占少量资源，不改变每 Core 1000 个可用 Node 的固定结论。
        if number % 30 == 1:
            kubernetes_pods.append({
                "apiVersion": "v1", "kind": "Pod",
                "metadata": {"name": f"bound-load-{number:04d}", "namespace": "ngd-ngg-test"},
                "spec": {
                    "nodeName": node_name,
                    "containers": [{
                        "name": "load", "image": "fixture.invalid/load:never-run",
                        "resources": {"requests": {"cpu": "1", "memory": "1Gi"}},
                    }],
                },
                "status": {"phase": "Running"},
            })

    static = {"clusterId": config["clusterId"], "topologyVersion": "dc-core-border-leaf-v1", "nodes": nodes}
    snapshot_id = canonical_hash(static)
    request_config = config["request"]
    request = {
        "requestId": "group1-3000-request",
        "taskUID": "task-group1-3000",
        "ngdUID": "ngd-group1-3000",
        "ngdGeneration": 1,
        "nodeStaticSnapshotId": snapshot_id,
        "podSets": [{
            "name": "distributed-workers",
            "replicas": request_config["replicas"],
            "minAvailable": request_config["minAvailable"],
            "resourcesPerPod": {"cpu": "1", "memory": "1Gi"},
        }],
        "nodeRequirements": {"nodeSelector": {"tests.ngg.io/worker": "true"}},
        "nodeUsageStates": states,
        "algorithms": [
            {"name": "requirement", "version": "v1", "parameters": {"requiredDistinctNodes": request_config["requiredDistinctNodes"]}},
            {"name": "topology", "version": "v1", "parameters": {"profile": "leaf-border-core-v1", "strategy": "NarrowestFit", "widestAllowedLevel": request_config["widestAllowedLevel"], "requiredDistinctNodes": request_config["requiredDistinctNodes"]}},
            {"name": "loadbalance", "version": "v1", "parameters": {"profile": "balanced-v2", "requireMetrics": True, "requireNetworkMetrics": True}},
        ],
        "maxCandidateGroups": request_config["maxCandidateGroups"],
    }
    return {
        "staticSnapshot": static,
        "snapshotId": snapshot_id,
        "dynamicState": {"scope": "request", "nodes": states},
        "metrics": metrics,
        "allocationRequest": request,
        "kubernetesNodes": kubernetes_nodes,
        "kubernetesPods": kubernetes_pods,
        "summary": {
            "nodeCount": len(nodes),
            "inUseCount": sum(1 for item in states if item["inUse"]),
            "availableCount": sum(1 for item in states if not item["inUse"]),
            "metricCount": len(METRIC_NAMES),
            "metricSamples": len(metrics) * len(METRIC_NAMES),
            "boundPodCount": len(kubernetes_pods),
            "coreCounts": {core: sum(1 for node in nodes if node["topology"]["coreSwitchId"] == core) for core in ("core-01", "core-02")},
        },
    }


def write_fixture(output: Path, data: dict[str, Any], include: set[str] | None = None) -> None:
    """只写当前测试组真正消费或需要展示的Fixture文件。"""

    output.mkdir(parents=True, exist_ok=True)
    files = {
        "node-static-snapshot.json": {"snapshotId": data["snapshotId"], **data["staticSnapshot"]},
        "node-dynamic-state.json": data["dynamicState"],
        "prometheus-metrics.json": {"source": "mock-prometheus-http", "nodes": data["metrics"]},
        "allocation-request.json": data["allocationRequest"],
        "kubernetes-nodes.json": {"items": data["kubernetesNodes"]},
        "kubernetes-pods.json": {"items": data["kubernetesPods"]},
        "fixture-summary.json": data["summary"],
    }
    selected = include if include is not None else set(files)
    unknown = selected.difference(files)
    if unknown:
        raise ValueError(f"unknown fixture files: {sorted(unknown)}")
    for name, value in files.items():
        if name in selected:
            write_json(output / name, value)


if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser(description="生成三组测试共用的确定性 3000 Node Fixture")
    parser.add_argument("--output", type=Path, required=True)
    parser.add_argument("--config", type=Path)
    args = parser.parse_args()
    value = generate(load_config(args.config))
    write_fixture(args.output, value)
    print(json.dumps({"output": str(args.output), "snapshotId": value["snapshotId"], **value["summary"]}, ensure_ascii=False))

File: go_test_suites/common/fixture.go
package common

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"time"

	corev1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/api/resource"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
	"k8s.io/apimachinery/pkg/types"
	"sigs.k8s.io/yaml"
)

// Fixture包含四组测试共用的确定性1000 Node输入。
type Fixture struct {
	NodeCount        int
	StaticSnapshot   map[string]any
	StaticSnapshotID string
	NodeUsageStates  []any
	Metrics          map[string]map[string]float64
	Allocation       map[string]any
	WorkerPayload    map[string]any
	Nodes            []corev1.Node
	Pods             []corev1.Pod
	Demand           *unstructured.Unstructured
}

// GenerateFixture用固定公式生成1000 Node、三层拓扑、动态状态和14项指标。
func GenerateFixture(nodeCount int) (*Fixture, error) {
	if nodeCount <= 0 || nodeCount%20 != 0 {
		return nil, fmt.Errorf("nodeCount must be a positive multiple of 20")
	}
	staticNodes := make([]any, 0, nodeCount)
	states := make([]any, 0, nodeCount)
	metrics := make(map[string]map[string]float64, nodeCount)
	nodes := make([]corev1.Node, 0, nodeCount)
	pods := []corev1.Pod{}
	leafCount := nodeCount / 20
	for number := 1; number <= nodeCount; number++ {
		leafNumber := (number-1)/20 + 1
		borderNumber := (leafNumber-1)*4/leafCount + 1
		coreNumber := (borderNumber-1)/2 + 1
		name := fmt.Sprintf("worker-%04d", number)
		uid := fmt.Sprintf("uid-worker-%04d", number)
		leaf := fmt.Sprintf("leaf-%03d", leafNumber)
		border := fmt.Sprintf("border-%02d", borderNumber)
		core := fmt.Sprintf("core-%02d", coreNumber)
		bandwidths := []float64{25, 40, 50, 100}
		latencies := []float64{4, 2.5, 1.5, 0.8}
		bandwidth := bandwidths[(leafNumber-1)%len(bandwidths)]
		latency := latencies[(leafNumber-1)%len(latencies)]
		labels := map[string]string{
			"tests.ngg.io/worker":                     "true",
			"topology.kubernetes.io/region":           "dc-test-01",
			"topology.kubernetes.io/rack":             fmt.Sprintf("rack-%03d", leafNumber),
			"topology.demo.ngg.io/topology-version":   "dc-core-border-leaf-v1",
			"topology.demo.ngg.io/core-switch":        core,
			"topology.demo.ngg.io/border-switch":      border,
			"topology.demo.ngg.io/leaf-switch":        leaf,
			"topology.demo.ngg.io/bandwidth-gbps":     fmt.Sprint(bandwidth),
			"topology.demo.ngg.io/latency-ms":         fmt.Sprint(latency),
			"topology.demo.ngg.io/convergence-switch": border,
		}
		topology := map[string]any{
			"dataCenter": "dc-test-01", "coreSwitchId": core, "borderSwitchId": border,
			"leafSwitchId": leaf, "switchId": leaf, "bandwidthGbps": bandwidth, "latencyMillis": latency,
		}
		staticNodes = append(staticNodes, map[string]any{
			"nodeName": name, "nodeUID": uid, "createdAt": "2026-08-21T00:00:00Z",
			"allocatable": map[string]any{"cpu": "32", "memory": "128Gi", "nvidia.com/gpu": "4"},
			"labels":      labels, "topology": topology,
		})
		inUse := number%3 == 0
		state := map[string]any{"nodeUID": uid, "inUse": inUse, "requestedResources": map[string]any{}}
		states = append(states, state)
		cpu := minFloat(0.95, 0.08+float64((leafNumber*13+number*7)%65)/100)
		memory := minFloat(0.95, 0.10+float64((leafNumber*11+number*5)%60)/100)
		receiveBPS := float64(20_000_000 + (number*7919)%600_000_000)
		transmitBPS := float64(15_000_000 + (number*6151)%500_000_000)
		receivePPS := float64(20_000 + (number*97)%800_000)
		transmitPPS := float64(18_000 + (number*89)%700_000)
		utilization := minFloat(0.95, (receiveBPS+transmitBPS)/(bandwidth*125_000_000))
		metrics[name] = map[string]float64{
			"cpuUsageRatio": cpu, "memoryUsageRatio": memory,
			"networkReceiveBytesPerSecond": receiveBPS, "networkTransmitBytesPerSecond": transmitBPS,
			"networkReceivePacketsPerSecond": receivePPS, "networkTransmitPacketsPerSecond": transmitPPS,
			"networkReceiveDropRatio":   float64(number%11) / 10000,
			"networkTransmitDropRatio":  float64(number%13) / 10000,
			"networkReceiveErrorRatio":  float64(number%7) / 20000,
			"networkTransmitErrorRatio": float64(number%5) / 20000,
			"tcpRetransmitRatio":        float64(number%17) / 10000,
			"networkLinkUpRatio": func() float64 {
				if number%101 == 0 {
					return 0.5
				}
				return 1
			}(),
			"networkUtilizationRatio":          utilization,
			"availableBandwidthBytesPerSecond": maxFloat(0, bandwidth*125_000_000-receiveBPS-transmitBPS),
		}
		allocatable := corev1.ResourceList{
			corev1.ResourceCPU: resource.MustParse("32"), corev1.ResourceMemory: resource.MustParse("128Gi"),
			corev1.ResourceName("nvidia.com/gpu"): resource.MustParse("4"),
		}
		nodes = append(nodes, corev1.Node{
			TypeMeta:   metav1.TypeMeta{APIVersion: "v1", Kind: "Node"},
			ObjectMeta: metav1.ObjectMeta{Name: name, UID: types.UID(uid), Labels: labels, CreationTimestamp: metav1.NewTime(time.Date(2026, 8, 21, 0, 0, 0, 0, time.UTC))},
			Spec:       corev1.NodeSpec{Unschedulable: inUse},
			Status:     corev1.NodeStatus{Capacity: allocatable.DeepCopy(), Allocatable: allocatable.DeepCopy(), Conditions: []corev1.NodeCondition{{Type: corev1.NodeReady, Status: corev1.ConditionTrue}}},
		})
		if number%100 == 1 {
			pods = append(pods, corev1.Pod{
				TypeMeta:   metav1.TypeMeta{APIVersion: "v1", Kind: "Pod"},
				ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("bound-load-%04d", number), Namespace: "ngd-ngg-test"},
				Spec:       corev1.PodSpec{NodeName: name, Containers: []corev1.Container{{Name: "load", Image: "fixture.invalid/load:never-run", Resources: corev1.ResourceRequirements{Requests: corev1.ResourceList{corev1.ResourceCPU: resource.MustParse("1"), corev1.ResourceMemory: resource.MustParse("1Gi")}}}}},
				Status:     corev1.PodStatus{Phase: corev1.PodRunning},
			})
		}
	}
	static := map[string]any{"clusterId": "mock-1000-node-cluster", "topologyVersion": "dc-core-border-leaf-v1", "nodes": staticNodes}
	snapshotID, err := CanonicalHash(static)
	if err != nil {
		return nil, err
	}
	ngdSpec := map[string]any{
		"schedulerName":       "volcano",
		"nodeSelector":        map[string]any{"matchLabels": map[string]any{"tests.ngg.io/worker": "true"}},
		"topologyRequirement": map[string]any{"profile": "leaf-border-core-v1", "strategy": "NarrowestFit", "widestAllowedLevel": "coreSwitch"},
		"maxCandidateGroups":  int64(3), "maxNodes": int64(6),
		"quota":        map[string]any{"cpu": "192", "memory": "768Gi"},
		"minResources": map[string]any{"cpu": "96", "memory": "384Gi"},
	}
	allocation := map[string]any{
		"requestId": "group-request-1000", "taskUID": "task-uid-1000", "ngdUID": "ngd-uid-1000", "ngdGeneration": int64(1),
		"requestMode": "resourcePool", "nodeStaticSnapshotId": snapshotID, "nodeUsageStates": states, "ngd": ngdSpec, "debugTrace": true,
	}
	metricSnapshotID, err := CanonicalHash(map[string]any{"catalogueVersion": "node-exporter-network-v1", "nodes": metrics})
	if err != nil {
		return nil, err
	}
	workerStatic := copyMap(static)
	workerStatic["snapshotId"] = snapshotID
	workerPayload := map[string]any{
		"request": allocation, "staticSnapshot": workerStatic,
		"metricSnapshot": map[string]any{
			"snapshotId": metricSnapshotID, "capturedAt": "2026-08-21T00:00:00Z", "capturedAtUnix": float64(1787270400),
			"catalogueVersion": "node-exporter-network-v1", "nodes": metrics,
		},
		"metricsDegraded": false, "warnings": []any{},
	}
	demand := &unstructured.Unstructured{Object: map[string]any{
		"apiVersion": "scheduling.platform.example.io/v1alpha1", "kind": "NodeGroupDemand",
		"metadata": map[string]any{"name": "go-test-demand", "labels": map[string]any{"scheduling.platform.example.io/scheduler": "volcano"}},
		"spec":     ngdSpec,
	}}
	return &Fixture{NodeCount: nodeCount, StaticSnapshot: static, StaticSnapshotID: snapshotID, NodeUsageStates: states, Metrics: metrics, Allocation: allocation, WorkerPayload: workerPayload, Nodes: nodes, Pods: pods, Demand: demand}, nil
}

// CanonicalHash与Algorithm静态缓存的JSON SHA-256规则一致。
func CanonicalHash(value any) (string, error) {
	raw, err := json.Marshal(value)
	if err != nil {
		return "", err
	}
	sum := sha256.Sum256(raw)
	return "sha256:" + hex.EncodeToString(sum[:]), nil
}

func WriteJSON(path string, value any) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	raw, err := json.MarshalIndent(value, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, append(raw, '\n'), 0o644)
}

func WriteYAML(path string, value any) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	raw, err := yaml.Marshal(value)
	if err != nil {
		return err
	}
	return os.WriteFile(path, raw, 0o644)
}

func ReadJSON(path string, target any) error {
	raw, err := os.ReadFile(path)
	if err != nil {
		return err
	}
	decoder := json.NewDecoder(bytes.NewReader(raw))
	decoder.UseNumber()
	return decoder.Decode(target)
}

// WriteFixtureInput将完整模拟输入写到指定组的testdata/input供现场查看。
func WriteFixtureInput(directory string, fixture *Fixture, includeKubernetes bool) error {
	static := copyMap(fixture.StaticSnapshot)
	static["snapshotId"] = fixture.StaticSnapshotID
	files := map[string]any{
		"node-static-snapshot.json": static,
		"node-dynamic-state.json":   map[string]any{"scope": "request", "nodes": fixture.NodeUsageStates},
		"prometheus-metrics.json":   map[string]any{"source": "mock-prometheus-http", "nodes": fixture.Metrics},
		"allocation-request.json":   fixture.Allocation,
		"worker-payload.json":       fixture.WorkerPayload,
	}
	if includeKubernetes {
		files["kubernetes-nodes.json"] = map[string]any{"items": fixture.Nodes}
		files["kubernetes-pods.json"] = map[string]any{"items": fixture.Pods}
		if err := WriteYAML(filepath.Join(directory, "ngd.yaml"), fixture.Demand.Object); err != nil {
			return err
		}
	}
	keys := make([]string, 0, len(files))
	for name := range files {
		keys = append(keys, name)
	}
	sort.Strings(keys)
	for _, name := range keys {
		if err := WriteJSON(filepath.Join(directory, name), files[name]); err != nil {
			return err
		}
	}
	return WriteJSON(filepath.Join(directory, "fixture-summary.json"), map[string]any{"nodeCount": fixture.NodeCount, "metricNamesPerNode": 14, "metricSamples": fixture.NodeCount * 14, "topology": "2 core / 4 border / 50 leaf / 20 nodes per leaf"})
}

func copyMap(source map[string]any) map[string]any {
	result := make(map[string]any, len(source))
	for key, value := range source {
		result[key] = value
	}
	return result
}

func minFloat(left, right float64) float64 {
	if left < right {
		return left
	}
	return right
}
func maxFloat(left, right float64) float64 {
	if left > right {
		return left
	}
	return right
}
